### DISCUSSION TITLE
Problem passing Nvidia GPU to Frigate on LXC

### DISCUSSION THREAD

**Discussion by btehan** [blakeblackshear/frigate#4019] [Ask A Question] [answered]
I'm probably running Frigate in the worst possible way, but I feel that I'm so close to getting this to work and must be missing something.  I'm running Frigate as a portainer stack on docker in a Debian LXC on proxmox 7.2.  I upgraded Frigate to 0.11.0-3846A13 earlier this week and it was working well.  Today I tried to pass through my Nvidia 1650 through to Frigate and it throws an error when updating the stack in portainer.

I've got the drivers setup on the host and in the LXC, nvidia-smi outputs the same on both:

> +-----------------------------------------------------------------------------+
> | NVIDIA-SMI 515.65.01    Driver Version: 515.65.01    CUDA Version: 11.7     |
> |-------------------------------+----------------------+----------------------+
> | GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
> | Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
> |                               |                      |               MIG M. |
> |===============================+======================+======================|
> |   0  NVIDIA GeForce ...  Off  | 00000000:10:00.0 Off |                  N/A |
> | 38%   36C    P0    11W / 100W |      0MiB /  4096MiB |      0%      Default |
> |                               |                      |                  N/A |
> +-------------------------------+----------------------+----------------------+
>                                                                                
> +-----------------------------------------------------------------------------+
> | Processes:                                                                  |
> |  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
> |        ID   ID                                                   Usage      |
> |=============================================================================|
> |  No running processes found                                                 |
> +-----------------------------------------------------------------------------+

I then added this to my frigate config file:

>ffmpeg:
>  hwaccel_args: -c:v h264_cuvid

And this to my portainer stack:

>    deploy:
>      resources:
>        reservations:
>          devices:
>            - driver: nvidia
>              count: 1
>              capabilities: [gpu]
>    environment:
>      - NVIDIA_VISIBLE_DEVICES=all
>      - NVIDIA_DRIVER_CAPABILITIES=all

Then it throws this error:

>Starting Error response from daemon: could not select device driver "nvidia" with capabilities: [[gpu]]

I've tweaked a number of things today, but here's my current stack:

>version: "3.9"
>services:
>  frigate:
>    container_name: frigate
>    privileged: true
>    restart: unless-stopped
>    image: blakeblackshear/frigate:stable
>    deploy:
>      resources:
>        reservations:
>          devices:
>            - driver: nvidia
>              count: 1
>              capabilities: [gpu]
>    shm_size: "64mb"
>    devices:
>      - /dev/apex_0:/dev/apex_0
>    volumes:
>      - /etc/localtime:/etc/localtime:ro
>      - /home/containers/frigate/config.yml:/config/config.yml:ro
>      - /home/containers/frigate:/db
>      - /mnt/tank/frigate:/media/frigate
>      - type: tmpfs
>        target: /tmp/cache
>        tmpfs:
>          size: 1000000000
>    ports:
>      - "5001:5000"
>      - "1935:1935"
>    environment:
>      - NVIDIA_VISIBLE_DEVICES=all
>      - NVIDIA_DRIVER_CAPABILITIES=all

Here's my current config.yml:

>mqtt:
>  host: 192.168.4.19
>  port: 1883
>  user: user
>  password: password
>
>database:
>  path: /db/frigate.db
>
>detectors:
>  coral1:
>    type: edgetpu
>    device: pci
>
>record:
>  enabled: True
>  retain:
>    days: 3
>    mode: motion
>  events:
>    retain:
>      default: 10
>      mode: active_objects
>
>detect:
>  stationary:
>    interval: 0
>    threshold: 50
>
>cameras:
>  driveway:
>    ffmpeg:
>      inputs:
>        - path: rtsp://user:password@192.168.4.99/cam/realmonitor?channel=1&subtype=1
>          roles:
>            - detect
>        - path: rtsp://user:password@192.168.4.99/cam/realmonitor?channel=1&subtype=0
>          roles: 
>            - record
>            - rtmp
>      hwaccel_args: -c:v h264_cuvid
>    detect:
>      width: 640
>      height: 480
>    snapshots:
>      enabled: True
>    objects:
>      track:
>        - person
>        - cat
>        - bicycle
>        - motorcycle
>        - bus
>        - dog
>        - car
>    zones:
>      driveway_zone:
>        coordinates: 0,332,60,252,124,201,192,134,226,130,262,130,419,130,484,146,640,209,640,480,0,480
>    motion:
>      mask:
>        - 640,172,421,129,184,125,111,125,43,125,0,135,0,0,640,0
>  doorbell:
>    ffmpeg:
>      inputs:
>        - path: rtsp://user:password@192.168.4.98/cam/realmonitor?channel=1&subtype=1
>          roles:
>            - detect
>        - path: rtsp://user:password@192.168.4.98/cam/realmonitor?channel=1&subtype=0
>          roles: 
>            - record
>            - rtmp
>      hwaccel_args: -c:v h264_cuvid  
>    detect:
>      width: 640
>      height: 480
>    snapshots:
>      enabled: True
>    objects:
>      track:
>        - person
>        - cat
>        - bicycle
>        - motorcycle
>        - bus
>        - dog
>        - car
>    zones:
>      front_porch:
>        coordinates: 640,480,640,350,64,325,64,480
>    motion:
>      mask:
>        - 0,480,64,480,64,327,166,325,517,353,640,350,640,0,0,0
>birdseye:
>  mode: continuous

And here's the config for the LXC container:

>arch: amd64
>cores: 4
>features: nesting=1
>hostname: docker
>memory: 8192
>mp0: /dev/bus/usb/007/002,mp:/dev/bus/usb
>net0: name=eth0,bridge=vmbr0,hwaddr=7E:4E:D9:D4:51:5F,ip=dhcp,type=veth
>onboot: 1
>ostype: debian
>rootfs: VM_LXC:subvol-107-disk-0,size=64G
>swap: 512
>lxc.cgroup2.devices.allow: a
>lxc.cgroup2.devices.allow: c 226:0 rwm
>lxc.cgroup2.devices.allow: c 226:128 rwm
>lxc.cgroup2.devices.allow: c 29:0 rwm
>lxc.cgroup2.devices.allow: c 189:* rwm
>lxc.cgroup2.devices.allow: c 120:* rwm
>lxc.cgroup.devices.allow: c 195:* rwm
>lxc.cgroup.devices.allow: c 234:* rwm
>lxc.cgroup.devices.allow: c 505:* rwm
>lxc.mount.entry: /dev/nvidia0 dev/nvidia0 none bind,optional,create=file
>lxc.mount.entry: /dev/nvidiactl dev/nvidiactl none bind,optional,create=file
>lxc.mount.entry: /dev/nvidia-uvm dev/nvidia-uvm none bind,optional,create=file
>lxc.mount.entry: /dev/nvidia-caps dev/nvidia-caps none bind,optional,create=file
>lxc.mount.entry: /dev/nvidia-uvm-tools dev/nvidia-uvm-tools none bind,optional,create=file
>lxc.apparmor.profile: unconfined
>lxc.mount.entry: /dev/apex_0 dev/apex_0 none bind,optional,create=file 0, 0
>lxc.mount.entry: /dev/dri/renderD128 dev/dri/renderD128 none bind,optional,create=file 0, 0
>lxc.cap.drop: 
>lxc.mount.auto: cgroup:rw
>lxc.autodev: 1
>lxc.hook.autodev: bash -c 'for char_dev in $(find /sys/dev/char -regextype sed  -regex ".*/1:1" -o -regex ".*/29:0" -o -regex ".*/188:.*" -o -regex ".*/189:.*" -o -regex ".*/226:0" -o -regex ".*/226:128"); do  dev="/dev/$(sed -n "/DEVNAME/ s/^.*=\(.*\)$/\1/p" ${char_dev}/uevent)";  mkdir -p $(dirname ${LXC_ROOTFS_MOUNT}${dev});  for link in $(udevadm info --query=property $dev | sed -n "s/DEVLINKS=//p"); do    mkdir -p ${LXC_ROOTFS_MOUNT}$(dirname $link);    cp -dpR $link ${LXC_ROOTFS_MOUNT}${link};  done;  cp -dpR $dev ${LXC_ROOTFS_MOUNT}${dev};done;'

When I comment out the deploy part of the stack and the hwaccel_args in the config.yml, frigate starts right up.

It is also possible I could be having multiple issues, however, as my path looks the same as issue [3223](https://github.com/blakeblackshear/frigate/issues/3223):

> # echo $PATH
> /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin

Thoughts on what I could try next?

**Comment by btehan** [comment, marked as answer]
Hokay, I've spun up a new LXC with docker inside and I've dumped portainer.  I'm using docker compose and going off the linked jellyfin documentation I decided to install nvidia-docker2.  That has made some progress, but it isn't exactly what I expected:

PVE host:
> root@pve:/etc/pve/lxc# nvidia-smi                    
> Sun Oct 16 21:05:40 2022       
> +-----------------------------------------------------------------------------+
> | NVIDIA-SMI 515.65.01    Driver Version: 515.65.01    CUDA Version: 11.7     |
> |-------------------------------+----------------------+----------------------+
> | GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
> | Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
> |                               |                      |               MIG M. |
> |===============================+======================+======================|
> |   0  NVIDIA GeForce ...  Off  | 00000000:10:00.0 Off |                  N/A |
> | 27%   37C    P2    16W / 100W |    159MiB /  4096MiB |      0%      Default |
> |                               |                      |                  N/A |
> +-------------------------------+----------------------+----------------------+
>                                                                                
> +-----------------------------------------------------------------------------+
> | Processes:                                                                  |
> |  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
> |        ID   ID                                                   Usage      |
> |=============================================================================|
> |    0   N/A  N/A   3835344      C   ffmpeg                             78MiB |
> |    0   N/A  N/A   3835349      C   ffmpeg                             78MiB |
> +-----------------------------------------------------------------------------+

LXC:
> root@docker:/home/containers#  nvidia-smi
> Mon Oct 17 02:06:53 2022       
> +-----------------------------------------------------------------------------+
> | NVIDIA-SMI 515.65.01    Driver Version: 515.65.01    CUDA Version: 11.7     |
> |-------------------------------+----------------------+----------------------+
> | GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
> | Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
> |                               |                      |               MIG M. |
> |===============================+======================+======================|
> |   0  NVIDIA GeForce ...  Off  | 00000000:10:00.0 Off |                  N/A |
> | 27%   38C    P2    16W / 100W |    159MiB /  4096MiB |      0%      Default |
> |                               |                      |                  N/A |
> +-------------------------------+----------------------+----------------------+
>                                                                                
> +-----------------------------------------------------------------------------+
> | Processes:                                                                  |
> |  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
> |        ID   ID                                                   Usage      |
> |=============================================================================|
> +-----------------------------------------------------------------------------+

Docker container:
> root@docker:/home/containers# docker exec -it frigate nvidia-smi
> Mon Oct 17 02:06:44 2022       
> +-----------------------------------------------------------------------------+
> | NVIDIA-SMI 515.65.01    Driver Version: 515.65.01    CUDA Version: 11.7     |
> |-------------------------------+----------------------+----------------------+
> | GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
> | Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
> |                               |                      |               MIG M. |
> |===============================+======================+======================|
> |   0  NVIDIA GeForce ...  Off  | 00000000:10:00.0 Off |                  N/A |
> | 27%   38C    P2    16W / 100W |    159MiB /  4096MiB |      0%      Default |
> |                               |                      |                  N/A |
> +-------------------------------+----------------------+----------------------+
>                                                                                
> +-----------------------------------------------------------------------------+
> | Processes:                                                                  |
> |  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
> |        ID   ID                                                   Usage      |
> |=============================================================================|
> +-----------------------------------------------------------------------------+

It seems strange to me that the output of the nvidia-smi command doesn't match up between the three.  I haven't noticed any difference in the performance of frigate.  Thoughts?

**Reply by NickM-27** [reply]
The container doesn't have the necessary permissions so the nvidia-smi output does not show processes. Given that the host is showing ffmpeg processes it appears that it is working as expected.

**Comment by btehan** [comment]
Yeah, I think you're right.  I ran nvidia-smi pmon on the host and it would output a scrolling list for the two ffmpeg  processes.  I then rebooted frigate from its GUI and nvidia-smi pmon immediately started outputting lines with no info on it:

`root@pve:~# nvidia-smi pmon
# gpu        pid  type    sm   mem   enc   dec   comman
# Idx          #   C/G     %     %     %     %   name
    0     821125     C     -     -     -     -   ffmpeg
    0    1314898     C     0     0     -     0   ffmpeg
    0     821125     C     -     -     -     -   ffmpeg
    0    1314898     C     0     0     -     0   ffmpeg
    0     821125     C     -     -     -     -   ffmpeg
    0    1314898     C     0     0     -     0   ffmpeg
    0     821125     C     0     0     -     0   ffmpeg
    0    1314898     C     0     0     -     0   ffmpeg
    0     821125     C     0     0     -     0   ffmpeg
    0    1314898     C     0     0     -     0   ffmpeg
    0     821125     C     -     -     -     -   ffmpeg
    0    1314898     C     0     0     -     0   ffmpeg
    0     821125     C     -     -     -     -   ffmpeg
    0    1314898     C     0     0     -     0   ffmpeg
    0     821125     C     -     -     -     -   ffmpeg
    0    1314898     C     0     0     -     0   ffmpeg
    0          -     -     -     -     -     -   -     
    0          -     -     -     -     -     -   -     
    0          -     -     -     -     -     -   -     
    0          -     -     -     -     -     -   -     
    0          -     -     -     -     -     -   -     
    0          -     -     -     -     -     -   -     
    0          -     -     -     -     -     -   -     
    0          -     -     -     -     -     -   -     
    0          -     -     -     -     -     -   -     
    0          -     -     -     -     -     -   -     
    0    1456854     C     0     0     -     0   ffmpeg
    0    1456875     C     -     -     -     -   ffmpeg
    0    1456854     C     0     0     -     0   ffmpeg
    0    1456875     C     -     -     -     -   ffmpeg
    0    1456854     C     0     0     -     0   ffmpeg
    0    1456875     C     -     -     -     -   ffmpeg
    0    1456854     C     0     0     -     0   ffmpeg
    0    1456875     C     -     -     -     -   ffmpeg`
